Set-Game/Set-Game/Game.swift

```diff
@@ -0,0 +1,193 @@
+//
+
+import Foundation
+
+struct Game {
+    
+    private var deck = CardDeck()
+    var deckCount: Int { return deck.cards.count }
+    
+    private(set) var score = 0
+    private(set) var iosScore = 0
+    private(set) var set = false
+    private(set) var playVersusIos = false
+    
+    private(set) var cardsOnTable = [Card]()
+    var cardsSelected = [Card]()
+    private(set) var cardsSets = [[Card]]()
+    private(set) var cardsHint = [Card]()
+    
+    private(set) var startTime = Date()
+    private(set) var endTime = Date()
+
+    
+    init() {
+        dealCards(12) { deal() }
+    }
+    
+    var isSet: Bool? {
+        get {
+            if cardsSelected.count == 3 {
+                return cardsSelected.isSet()
+            } else { return nil }
+        }
+        set {
+            if let newValue = newValue {
+                set = newValue
+                switch newValue {
+                case true:
+                    timer()
+                    cardsSets.append(cardsSelected)
+                    score += scoreBonus()
+                    replaceOrRemoveCard()
+                    cardsSelected.removeAll()
+                    startTime = Date()
+                case false:
+                    cardsSelected.removeAll()
+                    score += scorePenalty()
+                }
+            } else { cardsSelected.removeAll() }
+        }
+    }
+    
+    mutating func timer() {
+        endTime = Date()
+        let timeInterval: Double = endTime.timeIntervalSince(startTime)
+        
+        switch timeInterval {
+        case 0...7:
+            score += 2
+        case 7...15:
+            score += 1
+        default:
+            score += 0
+        }
+    }
+    
+    mutating func chooseCard(at index: Int) {
+        assert(cardsOnTable.indices.contains(index), "Concentration.chooseCard(at: \(index)): chosen index not in the cards")
+        
+        let chosenCard = cardsOnTable[index]
+        
+        if cardsSelected.contains(chosenCard), cardsSelected.count < 3{
+            cardsSelected = cardsSelected.filter() { $0 != chosenCard }
+        } else {
+            switch cardsSelected {
+            // if = 3 cards, check for SET match
+            case let cardsForSet where cardsForSet.count == 3:
+                isSet = isSet
+                cardsSelected.removeAll();
+                cardsSelected.append(chosenCard)
+            // if < 3 => add chosen card to selected
+            case let cardsForSet where !cardsForSet.contains(chosenCard):
+                cardsSelected.append(chosenCard)
+            // not a match, start building a set from currently chosen card
+            default: break
+            }
+        }
+    }
+    
+    mutating func reset() {
+        self = Game.init()
+    }
+    
+    mutating func dealThreeCards() {
+        dealCards(3) { deal() }
+    }
+    
+    private mutating func replaceOrRemoveCard() {
+        for cardSelected in cardsSelected {
```

Comment:
if you only need the index and not the object you can iterate directly through indices using cardsSelected.indices